ethereum/test/foundry/unit/concrete/Governance/Executing.t.sol

```diff
@@ -0,0 +1,225 @@
+            address(eventOnFallback),
+            0,
+            ""
+        );
+        governance.scheduleTransparent(validOp, 0);
+
+        IGovernance.Operation memory invalidOp = operationWithOneCallZeroSaltAndPredecessor(address(0), 0, "");
+        vm.expectRevert(bytes("Operation must be ready before execution"));
+        governance.execute(invalidOp);
+    }
+
+    function test_RevertWhen_ExecutingOperationWithDifferentValue() public {
+        vm.startPrank(owner);
+        IGovernance.Operation memory validOp = operationWithOneCallZeroSaltAndPredecessor(
+            address(eventOnFallback),
+            0,
+            ""
+        );
+        governance.scheduleTransparent(validOp, 0);
+
+        IGovernance.Operation memory invalidOp = operationWithOneCallZeroSaltAndPredecessor(
+            address(eventOnFallback),
+            1,
+            ""
+        );
+        vm.expectRevert(bytes("Operation must be ready before execution"));
+        governance.execute(invalidOp);
+    }
+
+    function test_RevertWhen_ExecutingOperationWithDifferentData() public {
+        vm.startPrank(owner);
+        IGovernance.Operation memory validOp = operationWithOneCallZeroSaltAndPredecessor(
+            address(eventOnFallback),
+            0,
+            ""
+        );
+        governance.scheduleTransparent(validOp, 0);
+
+        IGovernance.Operation memory invalidOp = operationWithOneCallZeroSaltAndPredecessor(
+            address(eventOnFallback),
+            0,
+            "00"
+        );
+        vm.expectRevert(bytes("Operation must be ready before execution"));
+        governance.execute(invalidOp);
+    }
+
+    function test_RevertWhen_ExecutingOperationWithDifferentPredecessor() public {
+        vm.startPrank(owner);
+        // Executing one operation to get a valid executed predecessor
+        IGovernance.Operation memory executedOp = operationWithOneCallZeroSaltAndPredecessor(
+            address(eventOnFallback),
+            0,
+            "1122"
+        );
+        governance.scheduleTransparent(executedOp, 0);
+        executeOpAndCheck(executedOp);
+
+        // Schedule & execute operation with 0 predecessor
+        IGovernance.Operation memory validOp = operationWithOneCallZeroSaltAndPredecessor(
+            address(eventOnFallback),
+            0,
+            ""
+        );
+        governance.scheduleTransparent(validOp, 0);
+        executeOpAndCheck(validOp);
+
+        // Schedule operation with predecessor of `executedOp` operation
+        IGovernance.Operation memory invalidOp = operationWithOneCallZeroSaltAndPredecessor(
+            address(eventOnFallback),
+            0,
+            ""
+        );
+        invalidOp.predecessor = governance.hashOperation(executedOp);
+
+        // Failed to execute operation that wasn't scheduled
+        vm.expectRevert(bytes("Operation must be ready before execution"));
+        governance.execute(invalidOp);
```

Comment:
I think for this test we need to stop `validOp` from being executed (and keep it only as scheduled). Otherwise even without overriding the predecessor this test would pass